test/jit/test_union.py

```diff
@@ -235,10 +233,47 @@ def fn2(x: Union[str, int]) -> str:
         self.assertEqual(fn1(1), fn2(1))
 
     def test_union_T_None_is_equivalent_to_optional_T(self):
-        def fn(x: Union[int, None]) -> str:
-            return "foo"
+        @torch.jit.script
+        def inner(x: Union[int, None]) -> int:
+            if x is not None:
+                return x
+            else:
+                return 5
 
-        self.checkScript(fn, (1,))
+        @torch.jit.script
+        def fn1() -> int:
+            a: Optional[int] = 5
+            b: Optional[int] = None
+            a_ = inner(a)
+            b_ = inner(b)
+            return a_ + b_
+
+        self.assertEqual(fn1(), 10)
+
+        @torch.jit.script
+        def inner2(x: Optional[int]) -> int:
+            if x is not None:
+                return x
+            else:
+                return 5
+
+        @torch.jit.script
+        def fn2() -> int:
+            a: Union[int, None] = 5
+            b: Union[int, None] = None
+            a_ = inner(a)
+            b_ = inner(b)
+            return a_ + b_
+
+        self.assertEqual(fn2(), 10)
+
+    def test_optional_of_union_is_flattened(self):
+        def fn() -> Union[int, str, None]:
```

Comment:
Current implementation can pass when one of the following is true 1) `Optional[Union]` is flatted 2) Value binding rule allows `Optional[Union]` to be assigned to a variable of type `Union`.

since we want to make sure 1 happens, we should rule out case 2. FileCheck is a good tool to use here to verify that in IR the `Optional` type is actually flattened.